server/src/logic/middleware/auth-validaton.middleware.ts

```diff
@@ -0,0 +1,129 @@
+          .addSource(AuthValidationMiddleware.name)
+          .addSourceMethod("checkRequest(..)");
+        return reject(e);
+      }
+
+      if (hasEmailProperty) {
+        req.body.isEmailLogin = true;
+      } else {
+        req.body.isEmailLogin = false;
+      }
+
+      resolve();
+    });
+  }
+
+  public extractRequestBody (req: Request): Promise<ExtractRequestBodyResult> {
+    return new Promise((resolve, reject) => {
+      if (!req.body.hasOwnProperty("isEmailLogin")) {
+        const e: ApiError = new ApiError(ERRORS.REQUEST_BODY)
+          .addSource(AuthValidationMiddleware.name)
+          .addSourceMethod("extractRequestBody(..)");
+        return reject(e);
+      }
+
+      req.body.ExtractRequestBodyResult = new ExtractRequestBodyResult(req.body.isEmailLogin,
+                                                (req.body.isEmailLogin ? req.body.email : req.body.username),
+                                                req.body.password);
+      resolve();
+    });
+  }
+
+  public checkValidToken (req: Request): Promise<any> {
+    return new Promise((resolve, reject) => {
+      const authHeader: string = (process.env.TOKEN_HEADER || Config.TOKEN_HEADER);
+      const token = req.body.token
+                    || req.query.token
+                    || req.headers[authHeader];
+      if (token) {
+        jwt.verify(token, (process.env.TOKEN_SECRET || Config.TOKEN_SECRET), (err: any, decoded: any)  => {
+          if (err) {
+            let e: ApiError;
+            if (err.name === "TokenExpiredError") {
+              e = new ApiError(ERRORS.AUTH_TOKEN_EXPIRED);
+            }
+            if (err.name === "JsonWebTokenError") {
+              e = new ApiError(ERRORS.AUTH_TOKEN_ERROR);
+            }
+            e.addSource(AuthValidationMiddleware.name).addSourceMethod("checkRequest(..)");
+            return reject(e);
+          }
+          req.body.decoded = decoded;
+          resolve();
+        });
+      } else {
+        const e = new ApiError(ERRORS.AUTH_TOKEN_MISSING)
+          .addSource(AuthValidationMiddleware.name)
+          .addSourceMethod("checkRequest(..)");
+        return reject(e);
+      }
+    });
+  }
+
+  public checkValidUser (req: Request): Promise<any> {
+    return new Promise((resolve, reject) => {
+      if (!req.body.hasOwnProperty("decoded")) {
+        const err: ApiError = new ApiError(ERRORS.REQUEST_BODY_AUTHCHECK.withArgs("a decoded Object"))
+          .addSource(AuthValidationMiddleware.name)
+          .addSourceMethod("checkValidUser(..)");
+        return reject(err);
+      }
+      this._userService.getUserByEmail(req.body.decoded.email)
+        .then((user: any) => {
+          delete req.body.decoded;
+          req.body.isAdmin = user.isadmin;
+          resolve();
+        })
+        .catch((error: any) => reject(error));
+    });
+  }
+
+  public checkUserRoleForRestriction (req: Request, shouldBeRestricted: boolean): Promise<any> {
```

Comment:
I will build this in a separate branch to show you what I mean.

I currently noticed that isAdmin is alway set to true. Is this a bug?